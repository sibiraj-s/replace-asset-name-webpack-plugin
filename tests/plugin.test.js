const fs = require('fs');
const path = require('path');

const getCompiler = require('./helpers/getCompiler');
const compile = require('./helpers/compile');

const ReplaceAssetNamePlugin = require('../src/index.js');

const getAssetNames = (stats) => {
  const { assets } = stats.compilation;
  return Object.keys(assets);
};

const hasAsset = (stats, assetName) => {
  const assets = getAssetNames(stats);
  if (assetName instanceof RegExp) {
    return assets.some((name) => assetName.test(name));
  }

  return assets.includes(assetName);
};

it('should replace the asset name correctly in dev mode', async () => {
  const compiler = getCompiler(true);

  new ReplaceAssetNamePlugin({
    test: /main.js$/,
    search: '{SECONDARY_FILE}',
    assetName: /secondary.js/,
  }).apply(compiler);

  const stats = await compile(compiler);

  const assets = getAssetNames(stats);
  expect(assets).toMatchSnapshot('assets');
  expect(assets.length).toBe(4);

  expect(hasAsset(stats, 'js/main.js')).toBeTruthy();

  const mainFilePath = path.resolve(process.cwd(), 'dist', 'js/main.js');

  const mainJS = await fs.promises.readFile(mainFilePath, 'utf-8');
  expect(mainJS).toContain('secondary.js');
});

it('should replace the asset name correctly in prod mode', async () => {
  const compiler = getCompiler();

  new ReplaceAssetNamePlugin({
    test: /main(?:-.*)?.js$/,
    search: '{SECONDARY_FILE}',
    assetName: /secondary(?:-.*)?.js$/,
  }).apply(compiler);

  const stats = await compile(compiler);

  const assets = getAssetNames(stats);
  expect(assets).toMatchSnapshot('assets');
  expect(assets.length).toBe(4);

  const mainFileName = assets.find((asset) => (/main(?:-.*)?.js$/).test(asset));
  const secondaryFileName = assets.find((asset) => (/secondary(?:-.*)?.js$/).test(asset));
  const mainFilePath = path.resolve(process.cwd(), 'dist', mainFileName);

  const mainJS = await fs.promises.readFile(mainFilePath, 'utf-8');
  expect(mainJS).toContain(secondaryFileName);
});

it('should do nothing when target asset is not found', async () => {
  const compiler = getCompiler();

  new ReplaceAssetNamePlugin({
    test: /main(?:-.*)?.js$/,
    search: '{SECONDARY_FILE}',
    assetName: /unknown.js$/,
  }).apply(compiler);

  const stats = await compile(compiler);

  const assets = getAssetNames(stats);
  expect(assets).toMatchSnapshot('assets');
  expect(assets.length).toBe(4);

  const mainFileName = assets.find((asset) => (/main(?:-.*)?.js$/).test(asset));
  const mainFilePath = path.resolve(process.cwd(), 'dist', mainFileName);

  expect(stats.toString()).toContain('no asset matches');

  const mainJS = await fs.promises.readFile(mainFilePath, 'utf-8');
  expect(mainJS).not.toContain('unknown.js');
});
