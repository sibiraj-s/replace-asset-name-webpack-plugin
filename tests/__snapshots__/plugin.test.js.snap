// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should do nothing when target asset is not found: assets 1`] = `
[
  "js/main-8d84ad0c7f998ea39855.js",
  "js/secondary-206f81d3a8599c10683b.js",
  "js/main-8d84ad0c7f998ea39855.js.map",
  "js/secondary-206f81d3a8599c10683b.js.map",
]
`;

exports[`should replace the asset name correctly in dev mode: assets 1`] = `
[
  "js/main.js",
  "js/secondary.js",
  "js/main.js.map",
  "js/secondary.js.map",
]
`;

exports[`should replace the asset name correctly in prod mode: assets 1`] = `
[
  "js/main-c4fe40039b79c556e104.js",
  "js/secondary-206f81d3a8599c10683b.js",
  "js/main-c4fe40039b79c556e104.js.map",
  "js/secondary-206f81d3a8599c10683b.js.map",
]
`;

exports[`should replace the asset name correctly when replace is a string: assets 1`] = `
[
  "js/main.js",
  "js/secondary.js",
  "js/main.js.map",
  "js/secondary.js.map",
]
`;

exports[`should replace the asset name correctly when search is a regex: assets 1`] = `
[
  "js/main-c4fe40039b79c556e104.js",
  "js/secondary-206f81d3a8599c10683b.js",
  "js/main-c4fe40039b79c556e104.js.map",
  "js/secondary-206f81d3a8599c10683b.js.map",
]
`;
